### ISSUE TITLE
Pin serde <=1.71.0 ? 

### ISSUE THREAD

**Issue by sanket1729** [rust-bitcoin/rust-bitcoin#2009]
TLDR: serde now ships with pre-compiled binary in an effort to reduce compile times. 

Related discussion: https://github.com/serde-rs/serde/issues/2538

Pinning seems to be easier way to go while the community figures out where this is going next.

**Comment by apoelstra** [comment]
Good find. We already have serde <1.171 (actually, at 1.156) in both our lockfiles for MSRV reasons. We should maybe update the README to link to the serde issue, or a blog post which I'm sure will be forthcoming, or whatever.

Meanwhile, serde itself has never broken us AFAIR ... perhaps this should finally be an impetus for us to purge `serde_derive` from our entire dependency tree. We can write our own simple macros in bitcoin-internals for common cases.

**Comment by sanket1729** [comment]
Agreed. We should purge serde derive and write our own macros for the derives.

**Comment by kayabaNerve** [comment]
1.0.185 removed the binaries. The sane thing would be to ban >=172, <185, if any banning is to occur.

**Comment by vincenzopalazzo** [comment]
>I think we should still drop serde_derive entirely.

I agree on this

**Comment by tcharding** [comment]
Post `servde v1.0.185` `serde_derive` is no longer an optional dependency of `serde`. Do we want to never update `serde` again and stick to pre-185?

I managed to get rid of `serde_derive` from "schemars"

**Comment by apoelstra** [comment]
> Post `servde v1.0.185` `serde_derive` is no longer an optional dependency of `serde`.

lol, wtf.

**Comment by apoelstra** [comment]
Yep, "working as intended" https://github.com/serde-rs/serde/issues/2601

I suppose we could just move `serde` itself to "second-class citizen" status and say there is no MSRV or guarantees of anything if you use the feature, it might download binary blobs, it might take 100x as long to compile, it might not compile at all, etc. And that versions after 1.0.185 are not tested by this project and if it's somehow broken it'll probably be a WONTFIX from us.

**Comment by tcharding** [comment]
Damn, that is a pretty tragic outcome from such a widely used crate as `serde`.

**Comment by apoelstra** [comment]
Yep. Though unsurprising, given that its wide use is the only reason it has survived so many rounds of broken-dependency pruning.

**Comment by kayabaNerve** [comment]
Is there data on how many users of bitcoin use the serde feature/what their use case is (just embedding into a larger serde object or some JSON transport)? I ask as someone who wasn't aware of its existence and never desired it (though I'm not the target user, obviously).

**Comment by tcharding** [comment]
TL;DR if we are at all serious about security do we really want enable software built with our library to use binary blobs (possibly quietly without devs knowing)? 

This binary blob thing is a pretty big deal, if we keep supporting `serde`, even with a warning that we only test against 1.0.85  it is highly likely that a user would just use latest `serde` and get the binary blob for `serder_derive` macros which then opens us up to all sorts of attacks. This could really come back to bite us years down the track when we have forgotten all about it.

**Comment by tcharding** [comment]
> Is there data on how many users of bitcoin use the serde feature/what their use case is (just embedding into a larger serde object or some JSON transport)? I ask as someone who wasn't aware of its existence and never desired it (though I'm not the target user, obviously).

We could just remove all the `serde` stuff an see if anyone complains.

**Comment by tcharding** [comment]
https://www.bleepingcomputer.com/news/security/rust-devs-push-back-as-serde-project-ships-precompiled-binaries/

**Comment by apoelstra** [comment]
@tcharding IIRC, dtolnay did relent and quietly removed the binary blob. So there were only a couple of affected serde reasons.

@kayabaNerve I use the serde support primarily to JSON-encode transactions and txouts. I use this for CLI input and on-disk storage for my wallet and my trading bot. JSON encoding is also heavily used by rust-bitcoincore-rpc and rust-jsonrpc, which is used to connect to Bitcoin Core and to various online API endpoints.

In theory we could directly support JSON encoding and decoding, but we would then need to pick a specific JSON library to support (or write our own, ugh) rather than letting users make their own decision.

**Comment by tcharding** [comment]
> We could just remove all the serde stuff an see if anyone complains.

Face palm, after reading @apoelstra's response this response of mine was so wrong, ouch.

**Comment by kayabaNerve** [comment]
@apoelstra That was what I'd expect to be the use-case, thanks for confirming.

Pinning to serde versions outside the binary blobs (or not pinning at all), and adding an extra warning on (lack of) support guarantees, sounds reasonable to me.

**Comment by vincenzopalazzo** [comment]
Wow I miss the party again 😸 

>I suppose we could just move serde itself to "second-class citizen" status and say there is no MSRV or guarantees of anything if you use the feature, it might download binary blobs, it might take 100x as long to compile, it might not compile at all, etc. And that versions after 1.0.185 are not tested by this project and if it's somehow broken it'll probably be a WONTFIX from us.

@apoelstra this looks like a good short-term solution to me, so concept ACK

>Is there data on how many users of bitcoin use the serde feature/what their use case is (just embedding into a larger serde object or some JSON transport)? I ask as someone who wasn't aware of its existence and never desired it (though I'm not the target user, obviously).

@kayabaNerve everyone is using serde, including the rust compiler itself :(

>We could just remove all the serde stuff and see if anyone complains.

I would love to draft a plan for this btw

>In theory we could directly support JSON encoding and decoding, but we would then need to pick a specific JSON library to support (or write our own, ugh) rather than letting users make their own decision.

This sounds like a very good long-term plan, if we remove Serde we remove also sync and quote and this means that we speed up the compilation time a lot. Now the question is if we create our JSON encoder we should use syn and quote as well, but I think I have the solution also for this https://github.com/rsmicro/kproc-macros

**Comment by apoelstra** [comment]
>This sounds like a very good long-term plan,

I'm pretty skeptical. There are multiple popular json encoders/decoders, all of which use serde and none of which we can compete with. We can pick one and implement direct conversions to its types, but when we'll have an indirect serde dependency.

I think the best we can do is keep the current situation where there is a direct but optional serde dependency.

Maybe we could do that **and** also find/write a direct JSON impl..

**Comment by vincenzopalazzo** [comment]
> We can pick one and implement direct conversions to its types, but when we'll have an indirect serde dependency.

Yeah this is also my concern, unfortunately side is tight with the ecosystem, but I will not consider a waste of time try it or provide also a second alternative

**Comment by kayabaNerve** [comment]
@vincenzopalazzo I mean bitcoin's serde feature, not serde as a crate. I'm well aware of how pervasive it is.

I only asked as I use the bitcoin binary encodings everywhere and was unsure how many parties used bitcoin types over JSON due to the lack of having any RPC code integrated Since apoelstra themselves can confirm their use, I'm fine dropping my question.

**Comment by jbesraa** [comment]
I am not sure if this https://docs.rs/tinyjson/latest/tinyjson meet all of rust-bitcoin requirements but its quite a well tested and documented crate. maybe we can clone it and build upon it in case we want to develop something internally

**Comment by vincenzopalazzo** [comment]
@jbesraa All these alternative libraries have a big problem, they are not compatible with serde_json::Value, so we have some problem I guess when user want to include this crate inside another application

**Comment by kayabaNerve** [comment]
miniserde also exists, which is by dtolnay. I believe tinyjson is a better candidate yet miniserde had yet to be mentioned.

**Comment by Kixunil** [comment]
I think this could be closed given the situation resolved?

Anyway, my 2c: I agree with @apoelstra that given popularity of `serde`, removing it is effectively impossible. As for removing derive, that might be feasible. I did so in my [project using proc macros](https://github.com/Kixunil/debcrafter/blob/master/src/input.rs#L17) because derive had horrible error messages. If we could somehow improve on error messages while we're at it that'd be fantastic. Which reminds me, we don't have `#[serde(deny_unknown_fields)]` on our types which is really, really not great.

**Comment by apoelstra** [comment]
Agreed, will close.
